### PR TITLE
Creating a repo from a template repo via API

### LINKED ISSUE #15934: FR: Creating a repo from a template repo via API
As title.

Needed to implement a CLI command (`tea repo create --template`) as in https://gitea.com/gitea/tea/issues/362.

### PR DESCRIPTION
**Pull request by a1012112796**
fix #15934
ref:
https://docs.github.com/en/rest/reference/repos#create-a-repository-using-a-template

### CODE REVIEW

**Review by zeripath — APPROVED** [review]

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -713,6 +713,7 @@ func Routes() *web.Route {
 				m.Combo("").Get(reqAnyRepoReader(), repo.Get).
 					Delete(reqToken(), reqOwner(), repo.Delete).
 					Patch(reqToken(), reqAdmin(), bind(api.EditRepoOption{}), repo.Edit)
+				m.Post("/generate", reqToken(), bind(api.GenerateRepoOption{}), repo.Generate)
```
Comment (lunny) [inline routers/api/v1/api.go]: The user should have permission to read the repository.

File: routers/api/v1/repo/repo.go [hunk 2]
```diff
@@ -307,6 +307,113 @@ func Create(ctx *context.APIContext) {
...
+		return
+	}
+
+	ctxUser := ctx.User
+	var err error
+	if form.Owner != ctxUser.Name {
+		ctxUser, err = models.GetOrgByName(form.Owner)
+		if err != nil {
+			if models.IsErrOrgNotExist(err) {
+				ctx.NotFound()
```
Comment (lunny) [inline routers/api/v1/repo/repo.go]: We should give an error message but not return 404.

**Comment by 6543** [comment]
love the new endpoint did had some inital work on this too but was not happy with it :)

**Comment by 6543** [comment]
@a1012112796 **TestAPIGenerateRepo** failed related

**Review by 6543 — APPROVED** [review]

**Pull request merged**